### PR TITLE
NEW: Add `reduce` function and tests

### PR DESCRIPTION
**Pull request by r-hyndman**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -54,6 +54,15 @@ function map(array, callback) {
...
 };
 
+function reduce(array, callback, initial) {
+  let accumulator = initial ?? array[0];
+  const startIndex = initial ? 0 : 1;
+  for (let i = startIndex; i < array.length; ++i) {
+    accumulator = callback(accumulator, array[i], i, array);
+  }
+  return accumulator;
+};
```
Comment (reubsinit) [inline src/index.js]: This is looking good, but there's something not quite right 🤔 . Compare the following invocation outputs:
```js
robDuce(data, sumOfValues, 0) // outputs 9

data.reduce(sumOfValues, 0) // outputs 10
```

File: tests/reduce.spec.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import {reduce} from '../src';
+
+/* eslint-disable require-jsdoc */
+const testArray = [1, 2, 3, 4];
+
+describe('reduce', () => {
```
Comment (reubsinit) [inline tests/reduce.spec.js:6]: I reckon it's worth including a test for executing `reduce` with an `initialValue` - see comment on `reduce` above

File: src/index.js [hunk 3]
```diff
@@ -54,6 +54,15 @@ function map(array, callback) {
   return result;
 };
 
+function reduce(array, callback, initial) {
+  let accumulator = initial ?? array[0];
+  const startIndex = initial ? 0 : 1;
```
Comment (reubsinit) [inline src/index.js]: This line is suspect 🔍 
```suggestion
  const startIndex = initial ? 0 : 1;
```

**Review by reubsinit — CHANGES_REQUESTED** [review]
Looking good, man. The problem with `reduce` is related to coercion.

File: src/index.js (hunk 1 above)
Comment (r-hyndman) [reply src/index.js]: robDuce 👌

File: tests/reduce.spec.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    const nativeResult = testArray.reduce(sumOfValues, 0);
+    const robDashResult = reduce(testArray, sumOfValues, 0);
+
+    // robDashResult should be the same value as nativeResult
+    expect(robDashResult).toEqual(nativeResult);
+  });
+
+  it('should return the highest value', () => {
+    function highestValue(accumulator, element) {
+      return accumulator || 0 > element ? accumulator : element;
```
Comment (reubsinit) [inline tests/reduce.spec.js]: What does this function do?

Comment (reubsinit) [reply tests/reduce.spec.js]: Test this out in runjs:
```js
const data = [1, 2, 7, 3, 4]

function reduce(array, callback, initial) {
  let accumulator = initial == undefined ? array[0] : initial;
  const startIndex = initial == undefined ? 1 : 0;
  for (let i = startIndex; i < array.length; ++i) {
    accumulator = callback(accumulator, array[i], i, array);
  }
  return accumulator;
};

function highestValue(accumulator, element) {
      return accumulator || 0 > element ? accumulator : element;
    }

data.reduce(highestValue)
reduce(data, highestValue)
```

**Review by reubsinit — CHANGES_REQUESTED** [review]
Looking good, Rob! I think there's something not working as expected in one of your specs.

File: tests/reduce.spec.js (hunk 4 above)
Comment (r-hyndman) [reply tests/reduce.spec.js]: Forgot the parentheses, oops. Should be:
```
function highestValue(accumulator, element) {
  return (accumulator || 0) > element ? accumulator : element;
}
```

Comment (reubsinit) [reply tests/reduce.spec.js]: What I'm wondering is why you need the optional check on `accumulator`? Why can't you just `return accumulator > element ? ...`?

My understanding is that `accumulator` should never be `undefined`.

Comment (r-hyndman) [reply tests/reduce.spec.js]: Yeah, that makes sense to me - I pulled that [function from this article](https://medium.com/poka-techblog/simplify-your-javascript-use-map-reduce-and-filter-bd02c593cc2d) that Al linked me a while back

**Review by reubsinit — APPROVED** [review]

**Pull request merged**